File: data/examples/python/bit_6.py
from typing import Tuple
from inline import Here

def _read_base128le(data: bytes) -> Tuple[int, int]:
    res = 0
    offset = 0
    while offset < len(data):
        o = data[offset]
        res += (o & 0x7F) << (7 * offset)
        Here().given(o, 0x70).given(offset, 0).given(res, 0).check_eq(res, 112)
        offset += 1
        if o < 0x80:
            # the Kaitai parser for protobuf support base128 le values up
            # to 8 groups (bytes). Due to the nature of the encoding, each
            # group attributes 7bit to the resulting value, which give
            # a 56 bit value at maximum.
            # The values which get encoded into protobuf variable length integers,
            # on the other hand, include full 64bit types (int64, uint64, sint64).
            # This means, the Kaitai encoder can not cover the full range of
            # possible values
            #
            # This decoder puts no limitation on the maximum value of variable
            # length integers. Values exceeding 64bit have to be handled externally
            return offset, res
    raise ValueError("varint exceeds bounds of provided data")

File: data/examples/java/bit_6.java
class Bit6 {
    public static long getShortDecimalValue(byte[] bytes) {
        long value = 0;
        if ((bytes[0] & 0x80) != 0) {
            for (int i = 0; i < 8 - bytes.length; ++i) {
                value |= 0xFFL << (8 * (7 - i));
                new Here().given(i, 0).given(value, 0).checkEq(value, -72057594037927936L);
            }
        }

        for (int i = 0; i < bytes.length; i++) {
            value |= ((long) bytes[bytes.length - i - 1] & 0xFFL) << (8 * i);
            new Here().given(i, 0).given(bytes, new byte[] { (byte) 0xe0, (byte) 0x4f }).given(value, 0).checkEq(value, 79L);
        }

        return value;
    }
}